### DISCUSSION TITLE
How to configure blog URLs?

### DISCUSSION THREAD

**Discussion by eli-oat** [datenstrom/community#239] [Ask a question] [answered]
By default yellow seems to use a pages title as its url slug, e.g. `title = about` generates the url website.com/about

Is there anyway to change a page's slug? 

I'm particularly interested in the context of blog pages. I'm looking to include the post's date in the url, e.g. website.com/2017/03/22/title

**Comment by BsNoSi** [comment]
What is the reason for this? I can not see any advantage over the current technique.

**Comment by eli-oat** [comment]
This is a standard url pattern for blog posts. See Jekyll, Wordpress, Cactus, etc. 

The primary reason I see for it being useful is that it helps prevent potential url collisions in larger blogs, especially those with multiple contributors.

**Comment by BsNoSi** [comment]
The fact that older systems use a similar approach for url addressing does not automatically make this method a better one.

Yellow adds by itself the date to the filename. If accidentally "same time, same title" happens, Yellow handles this by adding 2,3,4,… . This notifies to the author, that someone else has the same title on the same day – possibly a reason, to rethink it?

I consider it as a huge advantage that yellow *does not behave* as the enumerated systems.

**Comment by eli-oat** [comment]
To clarify — I'm not advocating for a wholesale change to yellow, just wondering if there is a way to support this different behavior.

**Comment by BsNoSi** [comment]
I didn't want to attack you with my comment – if that's what happened: Excuse me. 

I love the stringent addressing of Yellow. For me as a person, an address without many subdirectories is more readable and therefore more concise. The sub-directories have no (relevant) information value for me, an address tweeting with many subdirectories may then cut off the "exciting part", namely the title in the address.

Technically, there is (meanwhile) no need for a bunch of subdirectories. and if the technical limit is reached, it should be the job of the system, to handle that.

**Comment by markseuffert** [comment]
I think Yellow should offer this kind of flexibility. A system that can be adjusted to people's needs.

At the moment there are technical limitations that prevent you from using the date in the URL. The blog plugin only supports files in a single folder. If people would like to include more information in the URL, such as the publishing date, then plugins have to work across multiple folders and this means some redesign in Yellow's content handling. I can't promise anything, but will have a look what I can do.

If someone would like to see this feature, please add +1 to the top of this issue.

**Comment by eli-oat** [comment]
Thanks @markseu and @BsNoSi! 

I agree that the subdirectories are a clumsy solution to handling this. I am not 100% certain, but I think wordpress handles this sort of thing using a rewrite (granted, that use case is different to Yellow, since the content is being pulled from a database, rather than flat files). 

Rather than add subdirectories, e.g. `website.com/2017/03/29/content/` what about optionally appending the date to the content title for blog posts, e.g. `website.com/2017-03-29-content-title/`?

**Comment by markseuffert** [comment]
I don't mind if machines do more work, so that humans don't have to. 

Let's investigate the possibilities? I can think of postfixes, prefixes, subfolders and identifiers. Postfixes and prefixes are pretty common. Subfolders have both advantages and disadvantages, but they could come handy for all sorts of categorisations. With identifiers I mean that the CMS automatically adds a short unique ID.

From the following URL formats, which ones would you use?

1. website.com/article-date		: `website.com/made-for-people-2017-03-01`
2. website.com/date-article		: `website.com/2017-03-01-made-for-people`
3. website.com/date/article		: `website.com/2017-03-01/made-for-people`
4. website.com/year/article		: `website.com/2017/made-for-people`
5. website.com/article		&nbsp; &nbsp; &nbsp; &nbsp; &nbsp;: `website.com/made-for-people`
6. website.com/microblog	&nbsp; : `website.com/189129121`

**Comment by eli-oat** [comment]
For my use cases (I currently use Yellow for my [personal website](https://eli.li/), but I'm interested in starting to use it for client work, too), I think that 5 should remain the default, but I would certainly use options 4 or 3 if they were available. 

I think this could also be a great way to handle tags, `website.com/tag/article`

**Comment by BsNoSi** [comment]
`website.com/tag/article/` is not unique. How do you handle pages with more than one tag (what is quite usual)?

Something like `website.com/yyyy/mm/article` may be interesting for diaries.

Grouping, as currently done (`website.com/folder/article`) is close to `tag` grouping, possibly there is some subgrouping required (did not test this, I think, it should/would work).

**Comment by eli-oat** [comment]
@BsNoSi that is a good point about multiple tags. 

I've seen it work 2 ways in different places

1. A folder for each tag, with duplicate content across multiple folders
2. Each tag is nested in the previous, usually alphabetically. 

Neither option is perfect, though. Each has its issues.

**Comment by BsNoSi** [comment]
Duplicate content is critical: 

- How do you handle partial modifications – worst case scenario – of two editors at the same time? This requires a rights management overhead, that isn´t currently part of Yellow (and hopefully never ever will be implemented). 
- How do you handle the knowledge about duplicated content (equally same challenge as with two editors).

I think, it is cleaner and you can do it right away with Yellow, if you create folder („main tag“) with the content that matches there best and add your second, third,… tag, that will match on tag search (tag:xxx). This creates your „main tag folder adress“ without any requirement of copy and paste.

**Comment by markseuffert** [comment, marked as answer]
Just in time before Midsommar... more flexibility for blogs. The default behaviour is creating new pages in the same folder, as it has always been. But you can change this behaviour in your own blog. It's no problem if you would like to include publishing date in the URL. First you need to configure the blog settings, give it a quick test that your blog still works, then adjust the folder structure.

---

**Default - URL with blog folder**, e.g. `http://website/blog/made-for-people`

Use the default blog settings in file `system/extensions/yellow-system.ini`:

```
BlogStartLocation: auto
BlogNewLocation: @title
```

Adjust the `content` folder as following:

```
├── content
│   ├── 1-home
│   ├── 2-blog
│   │   ├── 2021-10-01-fika-is-good-for-you.md
│   │   ├── 2021-10-01-made-for-people.md
│   │   ├── 2021-10-11-blog-example-page.md
│   │   └── page.md
│   ├── 9-about
│   └── shared
├── media
└── system
```

---

**Example 1 - URL with full date**, e.g. `http://website/blog/2021-10-01/made-for-people`

Configure the following [blog settings](https://github.com/annaesvensson/yellow-blog#settings) in file `system/extensions/yellow-system.ini`:

```
BlogStartLocation: /blog/
BlogNewLocation: /blog/@date/@title
```

Adjust the `content` folder as following:

```
├── content
│   ├── 1-home
│   ├── 2-blog
│   │   ├── 2021-10-01
│   │   │   ├── 2021-10-01-fika-is-good-for-you.md
│   │   │   └── 2021-10-01-made-for-people.md
│   │   ├── 2021-10-11
│   │   │   └── 2021-10-11-blog-example-page.md
│   │   └── page.md
│   ├── 9-about
│   └── shared
├── media
└── system
```

---

**Example 2 - URL with subfolder for each year**, e.g. `http://website/blog/2021/made-for-people`

Configure the following [blog settings](https://github.com/annaesvensson/yellow-blog#settings) in file `system/extensions/yellow-system.ini`:

```
BlogStartLocation: /blog/
BlogNewLocation: /blog/@year/@title
```

Adjust the `content` folder as following:

```
├── content
│   ├── 1-home
│   ├── 2-blog
│   │   ├── 2020
│   │   │   └── 2020-06-22-fika-is-good-for-you.md
│   │   ├── 2021
│   │   │   ├── 2021-10-01-fika-is-good-for-you.md
│   │   │   ├── 2021-10-01-made-for-people.md
│   │   │   └── 2021-10-11-blog-example-page.md
│   │   └── page.md
│   ├── 9-about
│   └── shared
├── media
└── system
```

---

**Example 3 - URL with subfolder for each year and month**, e.g. `http://website/blog/2021/10/made-for-people`

Configure the following [blog settings](https://github.com/annaesvensson/yellow-blog#settings) in file `system/extensions/yellow-system.ini`:

```
BlogStartLocation: /blog/
BlogNewLocation: /blog/@year/@month/@title
```

Adjust the `content` folder as following:

```
├── content
│   ├── 1-home
│   ├── 2-blog
│   │   ├── 2020
│   │   │   └── 06
│   │   │       └── 2020-06-22-fika-is-good-for-you.md
│   │   ├── 2021
│   │   │   └── 10
│   │   │       ├── 2021-10-01-fika-is-good-for-you.md
│   │   │       ├── 2021-10-01-made-for-people.md
│   │   │       └── 2021-10-11-blog-example-page.md
│   │   └── page.md
│   ├── 9-about
│   └── shared
├── media
└── system
```

---

**Example 4 - Blog on home page, URL without subfolder**, e.g. `http://website/made-for-people`

Configure the following [blog settings](https://github.com/annaesvensson/yellow-blog#settings) in file `system/extensions/yellow-system.ini`:

```
BlogStartLocation: auto
BlogNewLocation: @title
```

Adjust the `content` folder as following:

```
├── content
│   ├── 1-blog
│   │   ├── 2021-10-01-fika-is-good-for-you.md
│   │   ├── 2021-10-01-made-for-people.md
│   │   ├── 2021-10-11-blog-example-page.md
│   │   └── page.md
│   ├── 9-about
│   └── shared
├── media
└── system
```

---

**Example 5 - Blog on home page, URL with subfolder for each year**, e.g. `http://website/2021/made-for-people`

Configure the following [blog settings](https://github.com/annaesvensson/yellow-blog#settings) in file `system/extensions/yellow-system.ini`:

```
BlogStartLocation: /
BlogNewLocation: /@year/@title
```

Adjust the `content` folder as following:

```
├── content
│   ├── 1-blog
│   │   └── page.md
│   ├── 9-about
│   ├── 2020
│   │   └── 2020-06-22-fika-is-good-for-you.md
│   ├── 2021
│   │   ├── 2021-10-01-fika-is-good-for-you.md
│   │   ├── 2021-10-01-made-for-people.md
│   │   └── 2021-10-11-blog-example-page.md
│   └── shared
├── media
└── system
```

---

**Example 6 - Blog on home page, URL with subfolder for each year and month**, e.g. `http://website/2021/10/made-for-people`

Configure the following [blog settings](https://github.com/annaesvensson/yellow-blog#settings) in file `system/extensions/yellow-system.ini`:

```
BlogStartLocation: /
BlogNewLocation: /@year/@month/@title
```

Adjust the `content` folder as following:

```
├── content
│   ├── 1-blog
│   │   └── page.md
│   ├── 9-about
│   ├── 2020
│   │   └── 06
│   │       └── 2020-06-22-fika-is-good-for-you.md
│   ├── 2021
│   │   └── 10
│   │       ├── 2021-10-01-fika-is-good-for-you.md
│   │       ├── 2021-10-01-made-for-people.md
│   │       └── 2021-10-11-blog-example-page.md
│   └── shared
├── media
└── system
```

---

Hope this is what you are looking for. Tested with Datenstrom Yellow 0.8.18.

**Reply by markseuffert** [reply]
Some tips for setting up the folder structure for blogs:

1. [Log in with your user account](https://github.com/annaesvensson/yellow-edit), go to `http://website/edit/blog/`
2. Configure the blog settings according to your needs, 
   click reload and give it a quick test that your blog still works
3. Add a new blog page, click `+` at the top right, then click `Create`

Now you have an example of how the `content` folder should look like. If you see a "file not found" error after structural changes on your website, it might help to reload the page or clear your browser cache. If you see a "file not found" error after deleting a folder on your website, go to the home page and then navigate to the blog start page.
